### PR TITLE
Allow assertions on the (root) cause message

### LINKED ISSUE #1776: Allow assertion on the root cause message
#### Summary

It is currently not directly possible to assert on the root cause message. There is a workaround possible by using `hasStackTraceContaining`, but it would be nice if this was supported directly.

See also https://stackoverflow.com/questions/38943346/assertj-assert-on-the-cause-message

#### Example

I would like to be able to do this:

```java
assertThatExceptionOfType(PersistenceException.class)
    .isThrownBy(() -> {
        // test code here
    })
   .withRootCauseInstanceOf(PSQLException.class)
   .withRootCauseMessageMatching("Key (user_id)=(.*) is not present in table \"book\"");
```

So to be consistent, probably the following new methods are needed:

* `withRootCauseMessageStartingWith(String description)`
* `withRootCauseMessageContaining(String description)` 
* `withRootCauseMessageContainingAll(CharSequence... values)`
* `withRootCauseMessageNotContaining(String content)`
* `withRootCauseMessageNotContainingAny(CharSequence... values)`
* `withRootCauseMessageMatching(String regex)`
* `withRootCauseMessageEndingWith(String description)`

### PR DESCRIPTION
**Pull request by wimdeblauwe**
#### Check List:
* Fixes #1776
* Unit tests : YES
* Javadoc with a code example (on API only) : YES

### CODE REVIEW

File: src/test/java/org/assertj/core/api/ThrowableAssertAlternative_havingRootCause_Test.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package org.assertj.core.api;
+
+import org.assertj.core.util.AssertionsUtil;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+@DisplayName("ThrowableAssertAlternative havingRootCause")
+class ThrowableAssertAlternative_havingRootCause_Test {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/ThrowableAssertAlternative_havingRootCause_Test.java]: same comments as `ThrowableAssertAlternative_havingCause_Test`

File: src/test/java/org/assertj/core/api/ThrowableAssertAlternative_havingCause_Test.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    //WHEN
+    ThrowableAssert.ThrowingCallable code = () -> {
+      Throwable throwable = new Throwable("top level message");
+      new ThrowableAssertAlternative<>(throwable)
+        .havingCause();
+    };
+    //THEN
+    AssertionsUtil.assertThatAssertionErrorIsThrownBy(code).withMessage("expecting java.lang.Throwable: top level message to have a cause but it did not");
+  }
+
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/ThrowableAssertAlternative_havingCause_Test.java]: add a test for when actual Throwable is null

File: src/test/java/org/assertj/core/api/ThrowableAssertAlternative_havingCause_Test.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  @Test
+  void should_fail_if_throwable_has_no_cause() {
+    //WHEN
+    ThrowableAssert.ThrowingCallable code = () -> {
+      Throwable throwable = new Throwable("top level message");
+      new ThrowableAssertAlternative<>(throwable)
+        .havingCause();
+    };
+    //THEN
+    AssertionsUtil.assertThatAssertionErrorIsThrownBy(code).withMessage("expecting java.lang.Throwable: top level message to have a cause but it did not");
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/ThrowableAssertAlternative_havingCause_Test.java]: this should build the error message with `shouldHaveCause`, here's an example from an `OptionalAssert` test
```java
  @Test
  public void should_fail_if_optional_is_empty() {
    // GIVEN
    String expectedValue = "something";
    // WHEN
    AssertionError error = expectAssertionError(() -> assertThat(Optional.empty()).contains(expectedValue));
    // THEN
    assertThat(error).hasMessage(shouldContain(expectedValue).create());
  }
```

File: src/main/java/org/assertj/core/error/ShouldHaveRootCause.java [hunk 4]
```diff
@@ -42,6 +42,10 @@ public static ErrorMessageFactory shouldHaveRootCause(Throwable actual, Throwabl
     return new ShouldHaveRootCause(actual, actualCause, expectedCause);
   }
 
+  public static ErrorMessageFactory shouldHaveRootCause(Throwable actualCause) {
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/error/ShouldHaveRootCause.java]: add a unit test in `ShouldHaveRootCause_create_Test`

File: src/main/java/org/assertj/core/error/ShouldHaveCause.java [hunk 5]
```diff
@@ -31,6 +31,10 @@ public static ErrorMessageFactory shouldHaveCause(Throwable actualCause, Throwab
     return new ShouldHaveCause(actualCause, expectedCause);
   }
 
+  public static ErrorMessageFactory shouldHaveCause(Throwable actualCause) {
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/error/ShouldHaveCause.java]: Add a unit test in a new  test class `ShouldHaveCause_create_Test`, take `ShouldHaveRootCause_create_Test` as an example (sorry `ShouldHaveCause_create_Test` should have already existed!)

File: src/main/java/org/assertj/core/api/AbstractThrowableAssert.java [hunk 6]
```diff
@@ -158,6 +158,30 @@ public SELF hasNoCause() {
...
   }
 
+  /**
+   * Verifies that the actual {@code Throwable} has a cause.
+   *
+   * @return this assertion object.
+   * @throws AssertionError if the actual {@code Throwable} is {@code null}.
+   * @throws AssertionError if the actual {@code Throwable} does not have a cause.
+   */
+  public SELF hasCause() {
```
Comment (scordio) [inline src/main/java/org/assertj/core/api/AbstractThrowableAssert.java]: What if we add `getCause()` instead, which checks if there is a cause and returns a new `ThrowableAssert`? This would be symmetrical with `havingCause()`, which could become:

```java
  public ThrowableAssertAlternative<?> havingCause() {
    AbstractThrowableAssert<?, ?> causeAssert = delegate.getCause();
    return new ThrowableAssertAlternative<>(causeAssert.actual);
  }
```

Same would be for `getRootCause()` instead of `hasRootCause()`.

I propose this option also because `hasCause()` would have a slightly different purpose compared to the existing `hasCause(Throwable)`, i.e., checking the cause existence vs. comparison with the input value.

File: src/test/java/org/assertj/core/api/ThrowableAssertAlternative_havingCause_Test.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    new ThrowableAssertAlternative<>(throwable)
+      .havingCause()
+      .withMessage("cause message");
+  }
+
+  @Test
+  void should_fail_if_throwable_has_no_cause() {
+    //WHEN
+    ThrowableAssert.ThrowingCallable code = () -> {
+      Throwable throwable = new Throwable("top level message");
```
[the remark this answers is not in the document]
Comment (wimdeblauwe) [reply src/test/java/org/assertj/core/api/ThrowableAssertAlternative_havingCause_Test.java]: Do you mean this?

```
  void should_fail_if_throwable_has_no_cause() {
    //GIVEN
    Throwable throwable = new Throwable("top level message");
    ThrowableAssertAlternative<Throwable> taa = new ThrowableAssertAlternative<>(throwable);
    //WHEN
    ThrowableAssert.ThrowingCallable code = taa::havingCause;
    //THEN
    AssertionsUtil.expectAssertionError(code);//.withMessage("expecting java.lang.Throwable: top level message to have a cause but it did not");
  }
```

But how do I check the assertion message?

[the remark this answers is not in the document]
Comment (wimdeblauwe) [reply src/test/java/org/assertj/core/api/ThrowableAssertAlternative_havingCause_Test.java]: Ignore the above, found it.

File: src/main/java/org/assertj/core/api/AbstractThrowableAssert.java (hunk 6 above)
Comment (wimdeblauwe) [reply src/main/java/org/assertj/core/api/AbstractThrowableAssert.java]: Would it not be better to _add_ the `getCause()` method in `AbstractThrowableAssert` on top? In that case, it would be like this:

```
  public ThrowableAssert getCause() {
    hasCause();
    return new ThrowableAssert(actual.getCause());
  }
```

And `havingCause` in `ThrowableAssertAlternative` becomes:

```
  public ThrowableAssertAlternative<?> havingCause() {
    ThrowableAssert causeAssert = delegate.getCause();
    return new ThrowableAssertAlternative<>(causeAssert.actual);
  }
```

I return a `ThrowableAssert` since that is the type I am using. Returning `AbstractThrowableAssert<?,?>` like you propose also works, but not sure why that would be better?

Comment (scordio) [reply src/main/java/org/assertj/core/api/AbstractThrowableAssert.java]: > Would it not be better to add the getCause() method in AbstractThrowableAssert on top?

Looks good to me!

> Returning AbstractThrowableAssert<?,?> like you propose also works, but not sure why that would be better?

IIRC this is a pattern used in AssertJ to allow subclasses to return a different assertion type. @joel-costigliola what do you think about this?

Comment (joel-costigliola) [reply src/main/java/org/assertj/core/api/AbstractThrowableAssert.java]: yes `getCause` in `AbstractThrowableAssert` would be better as `havingCause()` looks a bit like to `hasCause(cause)` which may confuse users.

```java
assertThat(e).getCause()
             .hasMessage("cause message")
             .getRootCause()
             .hasMessage("root cause message");
```

We can return `AbstractThrowableAssert<?,?>` in case someone wants their own `MyThrowableAssert` which only need to extend `AbstractThrowableAssert<?,?>` (to be fair I don't know if people are doing that or not)

Comment (wimdeblauwe) [reply src/main/java/org/assertj/core/api/AbstractThrowableAssert.java]: Ok. PR updated for this now.

Comment (joel-costigliola) [reply src/main/java/org/assertj/core/api/AbstractThrowableAssert.java]: @wimdeblauwe we also need to add `"getCause"` and `"getRootCause"`  to `SoftProxies.METHODS_CHANGING_THE_OBJECT_UNDER_TEST` for soft assertions to work properly.

A test must be added in `SoftAssertionsTest` like `map_soft_assertions_should_work_with_navigation_methods` but for throwable (name it `throwable_soft_assertions_should_work_with_navigation_methods`).

We want to prove in that test that after navigating to cause and root cause soft assertions work and capture properly all errors.

Comment (wimdeblauwe) [reply src/main/java/org/assertj/core/api/AbstractThrowableAssert.java]: Thanks for the tip. Test is added. It failed without updating `SoftProxies` and adding both methods made it work.

**Comment by wimdeblauwe** [comment]
I have addressed all remarks so far. I probably need to update documentation as well. Can you give me some pointers about that?

**Comment by joel-costigliola** [comment]
@wimdeblauwe yes that would be great!
You can create a PR on assertj/doc and add a _Checking cause and root cause_ section like `====== Checking the exception message` in  https://assertj.github.io/doc/#assertj-core-exception-assertions-reference (don't worry about the title style, https://github.com/assertj/doc/issues/33 is going to address it)

**Comment by joel-costigliola** [comment]
Now that we have `getCause` and `getRootCause`, @wimdeblauwe @scordio do you think we still need  `hasCause()` and `hasRootCause()`?

**Comment by wimdeblauwe** [comment]
> Now that we have `getCause` and `getRootCause`, @wimdeblauwe @scordio do you think we still need `hasCause()` and `hasRootCause()`?

I have no preference, but if you decide to remove it, I would make it private methods for clarity.

**Comment by wimdeblauwe** [comment]
Documentation PR can be found at https://github.com/assertj/doc/pull/34

**Comment by scordio** [comment]
> Now that we have `getCause` and `getRootCause`, @wimdeblauwe @scordio do you think we still need `hasCause()` and `hasRootCause()`?

I would keep `getCause()` and not `hasCause()`, or if we want to keep both I'd rename `hasCause(Throwable)` to something else to highlight the different behavior (e.g., `hasCauseLike(Throwable)` or `hasCauseSimilarTo(Throwable)`).

Same for `getRootCause()` vs `hasRootCause()`.

**Comment by joel-costigliola** [comment]
I'm currently integrating the PR, as part of it I have inlined `hasCause()` and `hasRootCause()`.

**Comment by joel-costigliola** [comment]
Integrated thanks @wimdeblauwe!
